[Server] Add cgroup awareness to resource reflection

aylei opened the pull request:
<!-- Describe the changes in this PR -->

followup:
- https://github.com/skypilot-org/skypilot/pull/4731#discussion_r1962889953
- https://github.com/skypilot-org/skypilot/pull/4737/files/b973f2b99b18ac0b749afec7a7a73cfc8190cfc4#r1959068720

Since #4731 has been merged, now API Server can run in limited resources more reliably. This PR adds cgroup awareness so in most cases users don't have to set the env vars and the server will detect the right resource it can use automatically.

This PR also corrects the worker number choice of uvicorn, which does not respect the `SKYPILOT_POD_CPU_CORE_LIMIT` env previously.

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

romilbhardwaj commented:
This is great @aylei! BTW instead of the envvars we have in #4731, any thoughts on using https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ to expose CPU/memory allocated in k8s?

aylei commented:
> This is great @aylei! BTW instead of the envvars we have in #4731, any thoughts on using https://kubernetes.io/docs/concepts/workloads/pods/downward-api/ to expose CPU/memory allocated in k8s?

These env vars can be set automatically using downward-api, via something like:

```
        - name: SKYPILOT_MEMORY_LIMIT
          valueFrom:
            resourceFieldRef:
              containerName: skypilot-api
              resource: limits.memory
```

Also, the value should be parsed to number GBs in skypilot.

In most cases cgroup reflection is sufficient. A hypothetical case that might need Downward API is that some advanced users want to set the "memory limit" of API server using container's `requests.memory` instead of `limits.memory` to do some QoS tricks.

Michaelvll commented on sky/utils/common_utils.py:
```diff
@@ -792,4 +789,69 @@ def get_mem_size_gb() -> float:
+            with open(period_path, 'r', encoding='utf-8') as f:
+                period = float(f.read().strip())
+            return quota / period if quota > 0 else None
+    except (IOError, ValueError):
```
Would there be `FileNotFoundError` exception?

aylei replied:
IOError is an alias of OSError, which covers `FileNotFoundError`

Michaelvll reviewed (commented):
Using the cgroup sounds good to me! Might worth checking the CPU count implementation in other libraries with cgroup, e.g. pylint: https://github.com/pylint-dev/pylint/blob/d17bb0631ce13b00a68319f31425ddf5a57719eb/pylint/lint/run.py#L40-L122

aylei commented:
> Using the cgroup sounds good to me! Might worth checking the CPU count implementation in other libraries with cgroup, e.g. pylint: https://github.com/pylint-dev/pylint/blob/d17bb0631ce13b00a68319f31425ddf5a57719eb/pylint/lint/run.py#L40-L122

Thanks @Michaelvll ! Just updated the pull request to put `sched_affinity` into account and added a comment about why we do not honor `cpu.shares`.

The UT was updated accordingly.

Michaelvll commented on sky/utils/common_utils.py:
```diff
@@ -808,4 +800,92 @@ def get_mem_size_gb() -> float:
+        The memory limit in bytes, or None if there is no limit in cgroups.
+    """
+    try:
+        path = ('/sys/fs/cgroup/memory.max' if _is_cgroup_v2() else
```
not sure if this is related: https://github.com/ray-project/ray/blob/7c8e1c80376035bc4eaecabcbc96ab04c28e47fe/python/ray/_private/memory_monitor.py#L122-L126

aylei replied:
looks like ray only handles cgroup v1. It may affect us if when launching a ray cluster on Kubernetes with cgroup v2, we need to test the impact and fix it in ray I think

Michaelvll reviewed (approved):
Thanks @aylei! LGTM with a minor comment
